tests/test_server.py

```diff
@@ -1,46 +1,65 @@
-import json
-
-import connexion
 import pytest
+import requests
 
-server = connexion.FlaskApp(__name__,
-                            specification_dir='../src/ratbag_emu/openapi/',
-                            debug=True)
-server.add_api('ratbag-emu.yaml',
-                options={"swagger_ui": True},
-                arguments={'title': 'ratbag-emu'},
-                strict_validation=True,
-                validate_responses=True)
 
+class Client(object):
+    def __init__(self, url='http://localhost:8080'):
+        self.url = url
 
-def test_permissions():
-    from hidtools.uhid import UHIDDevice
+    def get(self, path):
+        return requests.get(f'{self.url}{path}')
 
-    try:
-        assert UHIDDevice() is not None
-    except PermissionError:
-        pytest.exit('Not enough permissions to create UHID devices')
+    def post(self, path, data=None, json=None):
+        return requests.post(f'{self.url}{path}',
+                             data=data,
+                             json=json)
 
+    def delete(self, path):
+        return requests.delete(f'{self.url}{path}')
 
-@pytest.fixture(autouse=True)
-def handle_devices():
-    import threading
+    def put(self, path, data=None, json=None):
+        return requests.put(f'{self.url}{path}',
+                            data=data,
+                            json=json)
 
-    from . import ratbag_emu
 
-    import ratbag_emu.server
-    from ratbag_emu.device_handler import DeviceHandler
+def reload_udev_rules():
+    import subprocess
+
+    subprocess.run('udevadm control --reload-rules'.split())
+    subprocess.run('udevadm hwdb --update'.split())
+
+
+@pytest.fixture(scope='session', autouse=True)
```

Comment:
I *think* now that you're not using the flask wrapper anymore, this is sort of maybe obsolete. And on that note: you'd benefit from putting your tests into a `class TestSomething(unittest.TestCase)` with a `setup` and `teardown` method.

look at the hid-tools tests, they use that approach. You'd have one global setup for the udev rules and one class-specific one that can handle the `add_device` for you.